### PR TITLE
More Enchantment API

### PR DESCRIPTION
**Pull request by Machine-Maker**

### CODE REVIEW

File: Spigot-API-Patches/0295-More-Enchantment-API.patch [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
++    public abstract io.papermc.paper.enchantments.EnchantmentRarity getRarity();
++
++    /**
++     * Gets the damage increase as a result of the level and entity category specified
++     *
++     * @param level the level of enchantment
++     * @param entityCategory the category of entity
++     * @return the damage increase
++     */
++    public abstract float getDamageIncrease(int level, @NotNull org.bukkit.entity.EntityCategory entityCategory);
```
Comment (Proximyst) [inline Spigot-API-Patches/0295-More-Enchantment-API.patch]: This doesn't apply to all enchantments, though. Perhaps return a nullable Float instead?

File: Spigot-Server-Patches/0722-More-Enchantment-API.patch [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+ 
++    public Enchantment.Rarity getRarity() { return d(); } // Paper - OBFHELPER
+     public Enchantment.Rarity d() {
+         return this.d;
+     }
+@@ -70,6 +71,7 @@ public abstract class Enchantment {
+         return 0;
+     }
+ 
++    public float getDamangeIncrease(int level, EnumMonsterType enumMonsterType) { return a(level, enumMonsterType); } // Paper - OBFHELPER
```
Comment (Proximyst) [inline Spigot-Server-Patches/0722-More-Enchantment-API.patch]: damange

File: Spigot-Server-Patches/0722-More-Enchantment-API.patch [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
++            return EnumMonsterType.UNDEFINED;
++        } else if (entityCategory == EntityCategory.UNDEAD) {
++            return EnumMonsterType.UNDEAD;
++        } else if (entityCategory == EntityCategory.ARTHROPOD) {
++            return EnumMonsterType.ARTHROPOD;
++        } else if (entityCategory == EntityCategory.ILLAGER) {
++            return EnumMonsterType.ILLAGER;
++        } else if (entityCategory == EntityCategory.WATER) {
++            return EnumMonsterType.WATER_MOB;
++        }
```
Comment (Proximyst) [inline Spigot-Server-Patches/0722-More-Enchantment-API.patch]: https://www.w3schools.com/java/java_switch.asp

File: Spigot-Server-Patches/0722-More-Enchantment-API.patch [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
++
++        throw new IllegalArgumentException(String.format("%s is an unrecognized entity category", entityCategory));
```
Comment (Proximyst) [inline Spigot-Server-Patches/0722-More-Enchantment-API.patch]: Just use concatenation

File: Spigot-Server-Patches/0722-More-Enchantment-API.patch [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
++import net.minecraft.world.item.enchantment.Enchantment.Rarity;
++import org.bukkit.craftbukkit.enchantments.CraftEnchantment;
++import org.junit.Test;
++
++import static org.junit.Assert.assertNotNull;
++
++public class EnchantmentRarityTest {
++
++    @Test
++    public void test() {
```
Comment (Proximyst) [inline Spigot-Server-Patches/0722-More-Enchantment-API.patch]: Make this clear that its purpose is not to throw

File: Spigot-Server-Patches/0722-More-Enchantment-API.patch [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
++        for (Field field : EnumMonsterType.class.getDeclaredFields()) {
++            if (field.getType() == EnumMonsterType.class) {
++                enumMonsterTypeSet.add((EnumMonsterType) field.get(null));
++            }
++        }
++
++        for (EntityCategory entityCategory : EntityCategory.values()) {
++            enumMonsterTypeSet.remove(CraftLivingEntity.fromBukkitEntityCategory(entityCategory));
++        }
++        assertTrue(String.format("Found a %s instance that is missing a bukkit counterpart", EnumMonsterType.class), enumMonsterTypeSet.size() == 0);
```
Comment (Proximyst) [inline Spigot-Server-Patches/0722-More-Enchantment-API.patch]: Print them

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0295-More-Enchantment-API.patch (hunk 1 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0295-More-Enchantment-API.patch]: So if it’s <= 0 just return null? I just was returning what the nms abstract Enchantment returned. It defaults to 0, not null. Could note in the docs that this returns 0 for enchantments that don’t affect damage

File: Spigot-Server-Patches/0722-More-Enchantment-API.patch (hunk 6 above)
Comment (Machine-Maker) [reply Spigot-Server-Patches/0722-More-Enchantment-API.patch]: Print the ones left over? Is that even useful? It’s just gonna be EnumMonsterClass@276ae76 or smth

File: Spigot-API-Patches/0295-More-Enchantment-API.patch (hunk 1 above)
Comment (kashike) [reply Spigot-API-Patches/0295-More-Enchantment-API.patch]: I think `0f` makes more sense than `@Nullable Float` myself.

File: Spigot-Server-Patches/0722-More-Enchantment-API.patch [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+diff --git a/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java b/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java
+index 5a14430f63894bbe9daa42900cf5a6519bea4f45..c31f7864e0e1034b706a73f6752347a29212c125 100644
+--- a/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java
++++ b/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java
+@@ -71,7 +71,7 @@ public class CraftEnchantment extends Enchantment {
+ 
+     @Override
+     public boolean isCursed() {
+-        return target instanceof EnchantmentBinding || target instanceof EnchantmentVanishing;
++        return target.c(); // Paper
```
Comment (kashike) [inline Spigot-Server-Patches/0722-More-Enchantment-API.patch]:
```suggestion
+        return target.isCursed(); // Paper
```

**Review by kashike — DISMISSED** [review]

File: Spigot-Server-Patches/0722-More-Enchantment-API.patch (hunk 6 above)
Comment (Machine-Maker) [reply Spigot-Server-Patches/0722-More-Enchantment-API.patch]: Ok, I just changed it to a Map with the field name and printed that.

File: Spigot-Server-Patches/0722-More-Enchantment-API.patch (hunk 7 above)
Comment (Machine-Maker) [reply Spigot-Server-Patches/0722-More-Enchantment-API.patch]: done

File: Spigot-API-Patches/0295-More-Enchantment-API.patch (hunk 1 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0295-More-Enchantment-API.patch]: Ok, leaving it as `0f`

**Comment by Machine-Maker** [comment]
Should be all good for another review

File: Spigot-Server-Patches/0726-More-Enchantment-API.patch [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+diff --git a/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java b/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java
+index b76ef55145336cc8dc4857b79767f5a738ad5144..f17a4fa0fffd18d8e53eed9cf65a223db2b620ff 100644
+--- a/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java
++++ b/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java
+@@ -17,7 +17,7 @@ import net.minecraft.world.item.ItemStack;
+ 
+ public abstract class Enchantment {
+ 
+-    private final EnumItemSlot[] a;
++    private final EnumItemSlot[] a; public final EnumItemSlot[] getSlots() { return a; } // Paper - OBFHELPER
```
Comment (Proximyst) [inline Spigot-Server-Patches/0726-More-Enchantment-API.patch]:
```suggestion
+    private final EnumItemSlot[] a; public final EnumItemSlot[] getSlots() { return this.a; } // Paper - OBFHELPER
```
Be explicit to make Paperweight happy. Applies everywhere.

File: Spigot-API-Patches/0297-More-Enchantment-API.patch [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
++     */
++    public abstract float getDamageIncrease(int level, @NotNull org.bukkit.entity.EntityCategory entityCategory);
++
++    /**
++     * Gets the equipment slots where this enchantment is considered "active".
++     *
++     * @return the equipment slots
++     */
++    @NotNull
++    public abstract java.util.Set<org.bukkit.inventory.EquipmentSlot> getActiveSlots();
```
Comment (Proximyst) [inline Spigot-API-Patches/0297-More-Enchantment-API.patch]: Can't wait for people to be angry over these in their implementations 😆

**Review by Proximyst — APPROVED** [review]
LGTM. Note for merger.

File: Spigot-API-Patches/0297-More-Enchantment-API.patch (hunk 9 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0297-More-Enchantment-API.patch]: Should new methods not be abstract and just return null or 0 or an empty set or whatever and have them overriden in the wrapper and CraftEnchantment? I did kinda forget that people extend this class themselves in places so maybe not requiring it to be overriden would be better?

Comment (Proximyst) [reply Spigot-API-Patches/0297-More-Enchantment-API.patch]: No, absolutely not, they're fine as-is. Was mostly mocking folks who implement API classes that should never be extended, and they go out of their way to hacking in registrations.

Comment (Machine-Maker) [reply Spigot-API-Patches/0297-More-Enchantment-API.patch]: Yeah, that's what I figured, but thought I'd check to be sure.

**Review by kennytv — APPROVED** [review]
Thänk

**Pull request merged**